libs/designer/src/lib/ui/mcp/panel/operation/EditOperationPanelInner.tsx

```diff
@@ -17,62 +26,173 @@ export const EditOperationPanelInner = () => {
   const styles = useMcpPanelStyles();
   const [isDirty, setIsDirty] = useState(false);
 
+  const editOperationRef = useRef<EditOperationRef>(null);
+
+  const handleDismiss = useCallback(() => {
+    dispatch(closePanel());
+  }, [dispatch]);
+
+  const { selectedOperationId, inputParameters } = useSelector((state: RootState) => ({
+    selectedOperationId: state.connector.selectedOperationId,
+    inputParameters: state.operation.inputParameters,
+  }));
+
+  const parameters = selectedOperationId ? inputParameters[selectedOperationId] : null;
+
   const INTL_TEXT = {
     title: intl.formatMessage({
-      id: 'unVPy/',
-      defaultMessage: 'Edit Operations',
-      description: 'Title for edit operations panel',
+      id: 'KDsdC6',
+      defaultMessage: 'Edit Operation',
+      description: 'Title for edit operation panel',
     }),
     closeAriaLabel: intl.formatMessage({
       id: 'kdCuJZ',
       defaultMessage: 'Close panel',
       description: 'Aria label for close button',
     }),
+    unsavedChangesMessage: intl.formatMessage({
+      id: '2uxQXJ',
+      defaultMessage: 'You have unsaved changes. Are you sure you want to discard them?',
+      description: 'Confirmation message for discarding unsaved changes',
+    }),
   };
 
-  const handleDismiss = useCallback(() => {
-    dispatch(closePanel());
-  }, [dispatch]);
+  const resetToOriginalState = useCallback(() => {
+    editOperationRef.current?.resetLocalChanges();
+    setIsDirty(false);
+  }, []);
 
-  const updateOperationData = useCallback(
-    (_operationData: any) => {
-      setIsDirty(true);
-      //TODO: update operation data logic
-    },
-    [setIsDirty]
-  );
+  const handleCancel = useCallback(() => {
+    if (isDirty) {
+      const shouldDiscard = window.confirm(INTL_TEXT.unsavedChangesMessage);
+      if (!shouldDiscard) {
+        return;
+      }
+    }
+
+    resetToOriginalState();
+    handleDismiss();
+  }, [isDirty, INTL_TEXT.unsavedChangesMessage, resetToOriginalState, handleDismiss]);
+
+  const handleDismissWithReset = useCallback(() => {
+    resetToOriginalState();
+    handleDismiss();
+  }, [resetToOriginalState, handleDismiss]);
+
+  const handleSave = useCallback(() => {
+    if (!selectedOperationId || !parameters?.parameterGroups) {
+      LoggerService().log({
+        level: LogEntryLevel.Error,
+        message: 'Cannot save: missing operation data',
+        area: 'MCP_EditOperation',
+      });
+      return;
+    }
+
+    // Get all changes from the EditOperation component
+    const latestConditionalVisibilityChanges = editOperationRef.current?.getConditionalVisibilityChanges() || {};
+    const parameterValueChanges = editOperationRef.current?.getParameterValueChanges() || {};
+    const descriptionChange = editOperationRef.current?.getDescriptionChange();
+
+    Object.entries(latestConditionalVisibilityChanges).forEach(([parameterId, isVisible]) => {
+      let parameterGroupId: string | undefined;
```

Comment:
Use simple for loop its better on perf than these iterators

Reply:
I'm scanning every parameter group until you find a matching parameter, think we'll still need a double for loop; I guess we can break out early something like:
```
for (const [parameterId, isVisible] of Object.entries(latestConditionalVisibilityChanges)) {
  let parameterGroupId: string | undefined;

  for (const [groupId, group] of Object.entries(parameters.parameterGroups)) {
    if (group.parameters.find((p) => p.id === parameterId)) {
      parameterGroupId = groupId;
      break;
    }
  }

  if (parameterGroupId) {
    dispatch(
      updateParameterConditionalVisibility({
        nodeId: selectedOperationId,
        groupId: parameterGroupId,
        parameterId,
        value: isVisible,
      })
    );
  }
}
```